.circleci/config.yml

```diff
@@ -1,42 +1,217 @@
 # Ruby CircleCI 2.1 configuration file
 version: 2.1
 
-jobs:
-  build:
-    docker:
-      - image: cimg/ruby:3.0.2
-      - image: redis:6.2-alpine
+persist-to-workspace: &persist-to-workspace
+  persist_to_workspace:
+    root: /home/circleci
+    paths:
+      - .asdf
+      - .bashrc
+      - ./project
 
-    working_directory: ~/app
+attach-to-workspace: &attach-to-workspace
+  attach_workspace:
+    at: /home/circleci
 
-    steps:
-      - checkout
+executors:
+  ubuntu_vm:
+    machine:
+      image: ubuntu-2204:2024.04.4
+      docker_layer_caching: true
 
-      # Download and cache dependencies
-      - restore_cache:
+commands:
+  start_services:
+    parameters:
+      services:
+        description: List of services to run
+        type: string
+    steps:
+      - run:
+          name: Copy config files
+          command: |
+            cp script/config/podman-compose.yml ./compose.yaml
+            cp -r script/config/examples/sentinel1 ./sentinel1
+            cp -r script/config/examples/sentinel2 ./sentinel2
+            cp -r script/config/examples/sentinel3 ./sentinel3
+            cp -r script/config/examples/twemproxy.yml ./twemproxy.yml
+            cp -r script/config/examples/tls/master.conf ./master.conf
+            cp -r script/config/examples/tls/replica1.conf ./replica1.conf
+            cp -r script/config/examples/tls/replica2.conf ./replica2.conf
+            cp -r .circleci/ca-circleci-cert.pem ./ca-circleci-cert.pem
+            cp -r .circleci/circleci.crt ./circleci.crt
+            cp -r .circleci/circleci.key ./circleci.key
+            cp -r script/config/examples/tls/sentinel1 ./sentinel1_tls
+            cp -r script/config/examples/tls/sentinel2 ./sentinel2_tls
+            cp -r script/config/examples/tls/sentinel3 ./sentinel3_tls
+      - run:
+          name: Start services
+          command: |
+            docker-compose up << parameters.services >>
+          background: true
+  install_ruby:
+    steps:
+      - restore_cache: # Download and cache dependencies
           keys:
-          - bundle-{{ .Environment.CACHE_VERSION }}-{{ checksum "Gemfile.lock" }}
-
+            - asdf-ruby-3.0.2-v1
+      - run:
+          name: Install Ruby 3.0.2
+          command: |
+            cd $HOME
+            if [ ! -d ".asdf" ]; then
+              git clone https://github.com/asdf-vm/asdf.git ~/.asdf --branch v0.14.0
+            fi
+            echo '. "$HOME/.asdf/asdf.sh"' > ~/.bashrc
+            source .bashrc
+            asdf plugin add ruby https://github.com/asdf-vm/asdf-ruby.git
+            asdf install ruby 3.0.2
+            echo 'ruby 3.0.2' > $HOME/project/.tool-versions
+      - save_cache:
+          paths:
+            - ../.asdf
+          key: asdf-ruby-3.0.2-v1
```

Comment:
better use YAML anchor instead of hardcoding the key so that it can be changed in a single place later

Reply:
Done at https://github.com/3scale/apisonator/pull/385/commits/b1054977693e2509b970d67ab9812ba101327327